client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -219,4 +219,20 @@ private long getRemainingBytes(SegmentMetadataClientFactory metaFactory, StreamC
         return this.notifierFactory.getSegmentNotifier(this::createSynchronizer, executor);
     }
 
+    @Override
+    public Map<Stream, StreamCut> getStreamCuts() {
+        @Cleanup
+        StateSynchronizer<ReaderGroupState> synchronizer = createSynchronizer();
+        synchronizer.fetchUpdates();
+        ReaderGroupState state = synchronizer.getState();
+        Map<Stream, Map<Segment, Long>> positions = state.getPositions();
+        HashMap<Stream, StreamCut> cuts = new HashMap<>();
+
+        for (Entry<Stream, Map<Segment, Long>> streamPosition : positions.entrySet()) {
+            StreamCut position = new StreamCut(streamPosition.getKey(), streamPosition.getValue());
```

Comment:
Here you are passing the map pointer, not a copy of the map (second argument). In this case it looks harmless, but should we change the constructor of `StreamCut` to 
- Make a copy of the input map
- Make the map `unmodifiable` (since it's also exposing it via a getter)

Reply:
The constructor is currently being generated with the lombok @Data annotation. Should we do our own contractor and ditch the @Data annotation? It is also used elsewhere, like in the `unreadBytes` call in the same imply class. It might be better to do it separately. I can create an issue for it if you think it is worth it.

Reply:
In general, I think that we should make the change since it produces cleaner, less bug-prone code. If we create an instance of that class which takes in a map, then some other object either iterates or modifies it, we might run into trouble, not to mention cross thread safety.

I'd be ok with creating a separate issue for it, but we should at least try to get it done vs having it lost in the backlog

Reply:
State.getPostions() returns a newly allocated copy. So this class should be free to do with it as it sees fit and it will be safe by design. If you think it's appropriate we could rename or javadoc the method to make that more clear.

If should make a defensive copy of what is passed in is an interesting question. If we want to do it the way to go would be to have it construct an ImmutableMap from the passed map. This would also make the getter safe.